### PR TITLE
[#662] cli option to disable connection pool

### PR DESCRIPTION
**Pull request by usmansaleem**
## PR Description
Introduce cli option to disable internal database connection pooling library (i.e. Hikari). This can be used to use an external connection pool mechanism such as pgBouncer.

## Fixed Issue(s)

See #662
## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

**Review by siladu — COMMENTED** [review]
Reckon we should have a (acceptance?) test with connectionPool disabled?

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,15 @@
 - Better database pruning default values: Pruning enabled by default with `slashing-protection-pruning-epochs-to-keep = 250`, `slashing-protection-pruning-at-boot-enabled = false` and `slashing-protection-pruning-interval = 12`
 - Improved performance for slashing protection import
 
+### Experimental
+- Introduced cli option `--Xslashing-protection-db-connection-pool-enabled` to disable internal Hikari based connection
```
Comment (jframe) [inline CHANGELOG.md:13]: Why do we want this to be experimental?

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,15 @@
 - Better database pruning default values: Pruning enabled by default with `slashing-protection-pruning-epochs-to-keep = 250`, `slashing-protection-pruning-at-boot-enabled = false` and `slashing-protection-pruning-interval = 12`
 - Improved performance for slashing protection import
 
+### Experimental
+- Introduced cli option `--Xslashing-protection-db-connection-pool-enabled` to disable internal Hikari based connection 
+ pool to allow using external connection pool such as pgBouncer. `--slashing-protection-db-pool-configuration-file` and
```
Comment (jframe) [inline CHANGELOG.md:14]: /pool/database pool

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java [hunk 3]
```diff
@@ -115,6 +115,13 @@ public class PicoCliSlashingProtectionParameters implements SlashingProtectionPa
       arity = "1")
   private long dbHealthCheckIntervalMilliseconds = 30000;
 
+  @Option(
+      names = "--Xslashing-protection-db-connection-pool-enabled",
+      description =
+          "Set to false to disable internal database connection pooling. (Default: ${DEFAULT-VALUE})",
+      paramLabel = "<BOOL>")
+  private boolean connectionPoolEnabled = true;
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java:123]: nit: dbConnectionPoolEnabled to match the option name

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,15 @@
 - Better database pruning default values: Pruning enabled by default with `slashing-protection-pruning-epochs-to-keep = 250`, `slashing-protection-pruning-at-boot-enabled = false` and `slashing-protection-pruning-interval = 12`
 - Improved performance for slashing protection import
 
+### Experimental
+- Introduced cli option `--Xslashing-protection-db-connection-pool-enabled` to disable internal Hikari based connection 
+ pool to allow using external connection pool such as pgBouncer. `--slashing-protection-db-pool-configuration-file` and 
+ `--slashing-protection-db-pool-configuration-file` can be reused to specify PG Datasource properties.
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply CHANGELOG.md:15]: good catch. fixed.

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbConnection.java [hunk 5]
```diff
@@ -46,8 +48,14 @@ public static Jdbi createConnection(
...
       final String password,
-      final Path configurationFile) {
-    final DataSource datasource = createDataSource(jdbcUrl, username, password, configurationFile);
+      final Path configurationFile,
+      final boolean useCP) {
+    final DataSource datasource;
+    if (useCP) {
+      datasource = createHikariDataSource(jdbcUrl, username, password, configurationFile);
+    } else {
+      datasource = createPGDataSource(jdbcUrl, username, password, configurationFile);
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbConnection.java:57]: `PGSimpleDataSource` is a non-connection-pool datasource provided by PostgreSQL JDBC driver  ... so we can use any external connection pool (doesn't have to be pgBouncer). There is a connection-pool specific datasource provided by PostgreSQL JDBC driver as well, however, it is not recommended for production usage. Hikari, on the other hand, is a connection-pool specific datasource which can encapsulate any JDBC driver.

File: CHANGELOG.md (hunk 2 above)
Comment (usmansaleem) [reply CHANGELOG.md]: fixed.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java (hunk 3 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java]: refactored.

File: slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbConnection.java [hunk 6]
```diff
@@ -46,8 +48,14 @@ public static Jdbi createConnection(
       final String jdbcUrl,
       final String username,
       final String password,
-      final Path configurationFile) {
-    final DataSource datasource = createDataSource(jdbcUrl, username, password, configurationFile);
+      final Path configurationFile,
+      final boolean useCP) {
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbConnection.java]: refactored.

File: CHANGELOG.md [hunk 7]
```diff
@@ -9,6 +9,15 @@
 - Better database pruning default values: Pruning enabled by default with `slashing-protection-pruning-epochs-to-keep = 250`, `slashing-protection-pruning-at-boot-enabled = false` and `slashing-protection-pruning-interval = 12`
 - Improved performance for slashing protection import
 
+### Experimental
```
[the remark this answers is not in the document]
Comment (jframe) [reply CHANGELOG.md:12]: Think it is fine to include this in the changelog. It's clearly marked as experimental and could be useful for people trying out knowing the risks that it is experimental.

**Review by siladu — APPROVED** [review]
LGTM, it's just the changelog format change we discussed but not a blocker for me.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java [hunk 8]
```diff
@@ -115,6 +115,13 @@ public class PicoCliSlashingProtectionParameters implements SlashingProtectionPa
       arity = "1")
   private long dbHealthCheckIntervalMilliseconds = 30000;
 
+  @Option(
+      names = "--Xslashing-protection-db-connection-pool-enabled",
+      description =
+          "Set to false to disable internal database connection pooling. (Default: ${DEFAULT-VALUE})",
+      paramLabel = "<BOOL>")
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java:122]: Think should be hidden. Generally we hide experimental options

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java [hunk 9]
```diff
@@ -115,6 +115,13 @@ public class PicoCliSlashingProtectionParameters implements SlashingProtectionPa
       arity = "1")
   private long dbHealthCheckIntervalMilliseconds = 30000;
 
+  @Option(
+      names = "--Xslashing-protection-db-connection-pool-enabled",
+      description =
+          "Set to false to disable internal database connection pooling. (Default: ${DEFAULT-VALUE})",
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java:121]: Think we should add a note that this should only be done if external connection pooling is used.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java (hunk 8 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java]: added hidden=true.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java (hunk 9 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/PicoCliSlashingProtectionParameters.java]: Updated with the suggested note. (though the option is hidden!)

**Pull request merged**